SingularityService/src/test/java/com/hubspot/singularity/scheduler/SingularityRetryLostJobsTest.java

```diff
@@ -0,0 +1,77 @@
+package com.hubspot.singularity.scheduler;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.mesos.v1.Protos.TaskID;
+import org.apache.mesos.v1.Protos.TaskState;
+import org.apache.mesos.v1.Protos.TaskStatus;
+import org.apache.mesos.v1.Protos.TaskStatus.Reason;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+
+import com.google.inject.Inject;
+import com.hubspot.singularity.RequestType;
+import com.hubspot.singularity.SingularityPendingRequest.PendingType;
+import com.hubspot.singularity.SingularityTask;
+import com.hubspot.singularity.mesos.SingularityMesosStatusUpdateHandler;
+
+public class SingularityRetryLostJobsTest extends SingularitySchedulerTestBase {
+
+  // TODO: rm me
+  @Rule
+  public Timeout globalTimeout = new Timeout(10000, TimeUnit.SECONDS);
+
+  @Inject
+  SingularityMesosStatusUpdateHandler updateHandler;
+
+  public SingularityRetryLostJobsTest() {
+    super(false);
+  }
+
+  @Test
+  public void itRetriesLostShortRunningRequests() {
+    runTest(RequestType.ON_DEMAND, Reason.REASON_AGENT_RESTARTED, true);
+  }
+
+  @Test
+  public void itDoesNotRetryLostLongRunningRequests() {
+    runTest(RequestType.SERVICE, Reason.REASON_AGENT_RESTARTED, false);
+  }
+
+  @Test
+  public void itDoesNotRetryLostRequestsDueToNonAgentFailures() {
+    runTest(RequestType.ON_DEMAND, Reason.REASON_CONTAINER_LIMITATION_DISK, false);
+  }
+
+  private void runTest(RequestType requestType, Reason reason, boolean shouldRetry) {
+    initRequestWithType(requestType, false);
+    initFirstDeploy();
+
+    SingularityTask task = startTask(firstDeploy);
+    Assert.assertEquals(taskManager.getPendingTaskIds().size(), 0);
+    Assert.assertEquals(requestManager.getPendingRequests().size(), 0);
+    
+    try {
+      updateHandler.processStatusUpdateAsync(TaskStatus.newBuilder()
+          .setState(TaskState.TASK_LOST)
+          .setReason(reason)
+          .setTaskId(TaskID.newBuilder().setValue(task.getTaskId().getId()))
+          .build()).get();
+    } catch (InterruptedException | ExecutionException e) {
+      Assert.assertTrue(false);
+    }
+
+    if (shouldRetry) {
+      Assert.assertEquals(requestManager.getPendingRequests().size(), 1);
+      Assert.assertEquals(requestManager.getPendingRequests().get(0).getPendingType(), PendingType.RETRY);
+    } else {
+      System.out.println(requestManager.getPendingRequests());
+      Assert.assertEquals(requestManager.getPendingRequests().size(), 0);
+//      Assert.assertEquals(requestManager.getPendingRequests().get(0).getPendingType(), PendingType.TASK_DONE);
```

Comment:
I think I'm misunderstanding something because one of the two negative test cases fails here. The two negative test cases have different behaviors. One, `itDoesNotRetryLostRequestsDueToNonAgentFailures`, results in an empty pending queue, which is what I'd expect. The other, `itDoesNotRetryLostLongRunningRequests`, winds up with a request in the pending queue, even though I've manually verified through the debugger that the new "relaunch task" code path is never hit. This request also winds up with a `PendingType` of `TASK_DONE` which is definitely suspicious. Before I spend more time debugging, is there some obvious behavior I'm missing?

Reply:
So, that's normal that TASK_DONE is enqueued. What's different is that a TASK_DONE for a scheduled job won't necessarily result in a new task being started right away. We want to make sure we have our own pending request of a different type, that will trigger the desired behavior vs TASK_DONE which would do things like just schedule the next interval for a scheduled task instead of scheduling right away (and for an ON_DEMAND I believe TASK_DONE would be a no-op)